runtime/src/main/cpp/Memory.cpp

```diff
@@ -2252,6 +2253,96 @@ void FreezeSubgraph(ObjHeader* root) {
 #endif
 }
 
+void deepFrozenCopyTo(ObjHeader* to, ObjHeader* from, ObjHeaderMap* translationMapm, ObjHeader** firstBlocker);
+
+bool cannotClone(ObjHeader* original) {
+  if (original == nullptr)
+    return false;
+  if ((original->type_info()->flags_ & TF_NON_CLONEABLE) != 0)
+    return true;
+  if (original->has_meta_object() && ((original->meta_object()->flags_ & MF_NEVER_FROZEN) != 0))
+    return true;
+  return false;
+}
+
+OBJ_GETTER(makeShallowClone, ObjHeader* original) {
+  RuntimeAssert(!cannotClone(original), "Must be cloneable");
+  ObjHeader* clone;
+  if (IsArray(original)) {
+    clone = AllocArrayInstance(original->type_info(), original->array()->count_, OBJ_RESULT);
+    // Copy content only.
+    ::memcpy(clone->array() + 1, original->array() + 1, objectSize(original) - sizeof(ArrayHeader));
+  } else {
+    clone = AllocInstance(original->type_info(), OBJ_RESULT);
+    // Copy content only.
+    ::memcpy(clone + 1, original + 1, objectSize(original) - sizeof(ObjHeader));
+  }
+  RETURN_OBJ(clone);
+}
+
+void changeToCopyIfNeeded(ObjHeader** location, ObjHeaderMap* translationMap, ObjHeader** firstBlocker) {
```

Comment:
The name doesn't reflect the semantics with respect to all the complexities. Maybe `patchRawCopiedReference`?

Reply:
Not sure if it's dramatically better.